tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Grpc/GrpcLegacy/Client/MetadataArraySafeHandleCreateInstrumentation.cs

```diff
@@ -0,0 +1,158 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.ComponentModel;
+using Datadog.Trace.ClrProfiler.CallTarget;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.DuckTyping;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Grpc.GrpcLegacy.Client
+{
+    /// <summary>
+    /// Grpc.Core.Internal calltarget instrumentation
+    /// </summary>
+    [InstrumentMethod(
+        AssemblyName = "Grpc.Core",
+        TypeName = "Grpc.Core.Internal.MetadataArraySafeHandle",
+        MethodName = "Create",
+        ReturnTypeName = "Grpc.Core.Internal.MetadataArraySafeHandle",
+        ParameterTypeNames = new[] { "Grpc.Core.Metadata" },
+        MinimumVersion = "2.0.0",
+        MaximumVersion = "2.*.*",
+        IntegrationName = nameof(Grpc))]
+    [Browsable(false)]
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public static class MetadataArraySafeHandleCreateInstrumentation
+    {
+        /// <summary>
+        /// OnMethodBegin callback
+        /// </summary>
+        internal static CallTargetState OnMethodBegin<TInstance, TMetadata>(TInstance instance, TMetadata metadataInstance)
+        {
+            var tracer = Tracer.Instance;
+            if (!GrpcCoreApiVersionHelper.IsSupported || !tracer.Settings.IsIntegrationEnabled(IntegrationId.Grpc))
+            {
+                return CallTargetState.GetDefault();
+            }
+
+            // This integration is called in both the server and client-side paths
+            // and is responsible for serializing the metadata. However, we don't want it it serialize
+            // our Temporary headers, so we remove them before the serialize, and restore them afterwards.
+            // But we only add the extra headers for the client side code, so do a short-circuit check
+            // for one of the temporary headers that should always be there in client side code
+            var metadata = metadataInstance.DuckCast<IMetadata>();
+            if (GetAndRemove(metadata, TemporaryHeaders.Service) is { } service)
+            {
+                // Remove our temporary headers so they don't get sent over the wire
+                var methodKind = GetAndRemove(metadata, TemporaryHeaders.MethodKind);
+                var methodName = GetAndRemove(metadata, TemporaryHeaders.MethodName);
+                var startTime = GetAndRemove(metadata, TemporaryHeaders.StartTime);
+
+                var parentId = GetAndRemove(metadata, TemporaryHeaders.ParentId);
+                var parentService = GetAndRemove(metadata, TemporaryHeaders.ParentService);
+
+                if (methodKind is not null
+                 && methodName is not null
+                 && startTime is not null)
+                {
+                    metadata.Remove(methodKind);
+                    metadata.Remove(methodName);
+                    metadata.Remove(startTime);
```

Comment:
Are these calls still necessary?

Reply:
Oops, no, thanks! Fixed in https://github.com/DataDog/dd-trace-dotnet/pull/2545/commits/2cc39b45b524365b7edcf29c31c02045b393efcc